### PR TITLE
Development environment and semicolon linting fix

### PR DESCRIPTION
**Pull request by lukeburpee**
### Problem

The generator doesn't provide a development environment for typescript generated applications. This pull request provides a development script and nodemon configuration file to restart the ts-node server on source file change, with hot linting/type checking. 

This pull request also addresses linting errors that occur in typescript applications when the generator updates an existing application with a different semicolon configuration. Rather than providing a static tslint.json, the tslint.json file is generated and updated similarly to the .eslintrc file. Additionally, the write configurations are changed for the channels and log hook files to allow for update on semicolon option change.

### Issues

No open issues

### Dependencies

No dependencies

### CODE REVIEW

File: generators/writing/templates/nodemon.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+	"watch": ["src/**/*.ts"],
+	"execMap": {
+		"ts": "tslint -p tsconfig.json -c tslint.json && ts-node --files --seed --typeCheck"
```
Comment (j2L4e) [inline generators/writing/templates/nodemon.json]: wouldn't this reseed the db on every code change?

Comment (eddyystop) [reply generators/writing/templates/nodemon.json]: It wouldn't without NODE_ENV being set beforehand, either in the environment or in the command itself.

The generated scripts produce dual scripts such as `script` and `script:seed`.

**Comment by eddyystop** [comment]
@j2L4e Am I correct in remembering we might have had some issues with nodemon and TS? Or was it some other Node restart utility.

**Comment by j2L4e** [comment]
I was having trouble with file watchers in general at the time. I tried it and this seems to be working for me :+1:

**Comment by eddyystop** [comment]
We need to make a decision on this PR.

I would suggest --seed is removed because
- I'm uncomfortable it could be activated by a NODE_ENV that's not in the command itself.
- I think its a good idea for the data to not change if you are tracking down an issue caused by a combination of factors.
- Its easy to run `generate fakes` when you want to.

@lukeburpee  @j2L4e what do you think?

**Comment by eddyystop** [comment]
Let me know if you're interested in adding a `dev:seed` commnad to this PR. That would make it similar to the `start` script.

**Comment by lukeburpee** [comment]
Yep, I'd be up for that.

**Comment by eddyystop** [comment]
@lukeburpee Let me know when you have finished testing the changes and are sure the PR is 100% complete.

**Comment by lukeburpee** [comment]
@eddyystop and @j2L4e As a status update, 

- the dev and dev:seed scripts work for both ts and js (with automatic linting for js), but the current seed scripts re-seeds the database on reload (as expected). One potential fix that might work would be to store some boolean value in the generated seed data (e.g. "seed-updated": true/false), where the database would only be updated if the boolean where set to false. Re-running generate fakes would update the boolean to false, and the boolean value would be updated to true after the database has been seeded with the updated data. Of course, this could be saved for a separate pull request.

- While testing, I've gone through and corrected various lint and type errors that were uncovered, particularly concerning the seed-data and type declaration files. These corrections were based on the current tsconfig and tslint configurations. 

- I've added declarations for json imports and feathers-plus/test-utils when seeding is enabled, but not sure if this is how you want approach json imports in the future. Either way, it's only enabled when seeding is enabled.

- So far I've tested seeding for nedb and mongoose and both work as expected. The current mongodb.ts throws type errors, some of which have been fixed by adding the mongodb typings to devDependencies for mongodb services. Shouldn't be too difficult to get working.

File: generators/writing/templates/src/typings.d.ejs [hunk 2]
```diff
@@ -56,12 +53,21 @@ declare module 'feathers-<%- name %>'<%- sc %>
...
 // GraphQL.
 declare module '@feathers-plus/graphql'<%- sc %>
 declare module 'merge-graphql-schemas'<%- sc %>
 <% } -%>
 <%# --- if-5 ends above. -%>
+<%# --- if-7 starts below. -%>
+<% if (specs.app.seedData) { -%>
+// Seed.
+declare module '@feathers-plus/test-utils'<%- sc %>
+declare module '*.json' {
```
[the remark this answers is not in the document]
Comment (eddyystop) [reply generators/writing/templates/src/typings.d.ejs]: "One potential fix that might work would be to store some boolean value in the generated seed"

I re-seed my DB often as it gets mutated during manual testing and I want a stable starting point. So a flag preventing seeding more than once would be problematic. It also increases the cognative load on people.

IMHO persisted flags don't work well in general as they can end up in the wrong state. They then has to be reset manually and that similarly increases the cognative load.

Can The xxx be hacked to execute one command on initial startup and another on restarts?

**Comment by eddyystop** [comment]
This PR has gotten so big its scary. I hope we don't have too many merge conflicts.

- We need to stop adding to this PR, delaying more changes to another PR once this is merged.
- We need to decide on what if anything to do with nodemon and --seed
- @j2L4e needs to review the changes. That'll be a challenge.
- @j2L4e need to test the changes without publishing, yet being able to revert if necessary.

Its a bunch of work. How's your availability Jan?

**Comment by j2L4e** [comment]
Re nodemon: We could introduce another flag that terminates the process after seeding the db and then use something like `npm start seedonce && npm start dev`

I'm totally tied up until the end of next week, but I'll be able to test the changes then.

@lukeburpee Thanks a lot for your contribution! As @eddyystop mentioned, it'd be great if you split up future PRs so they can be reviewed / cherry-picked more easily.

**Comment by eddyystop** [comment]
I would still prefer a solution that did not depend on a persisted flag in some DB table.

**Comment by eddyystop** [comment]
@j2L4e I can fix all the tests once you've tested the PR. That'll take going on to 4 hrs.

You may find it helpful to at least run one or two of the TS source compares. You can limit the tests run by changing https://github.com/feathers-plus/generator-feathers-plus/blob/master/test/z-generators-writing.test.js#L1088-L1089

**Comment by eddyystop** [comment]
@lukeburpee Would you mind contacting on the Feathers Slack channel? Thanks. https://feathersjs.slack.com/messages/C52QSFK0A/

**Comment by NickBolles** [comment]
This closes #102, of course I only saw this after I made some changes for a PR... Looks awesome good job @lukeburpee! When will this be released?

**Comment by NickBolles** [comment]
@lukeburpee actually a couple suggestions around the config JSON files

First, could you include the `resolveJsonModule` in the tsconfig 

Also, does this fix: https://github.com/feathers-plus/generator-feathers-plus/issues/102#issuecomment-438745568?

**Comment by NickBolles** [comment]
I didn't see Eddystop's [comment](https://github.com/feathers-plus/generator-feathers-plus/pull/83#issuecomment-430350859) before. I can follow up with a PR for some of the changes I suggested after this is merged.

**Comment by eddyystop** [comment]
This PR will have to be broken into multiple smaller ones based on the current generator source. Otherwise I expect we will have a zoo of merge conflicts.

@j2L4e ^^^

**Comment by j2L4e** [comment]
@lukeburpee sorry for taking so long to review this! I second @eddyystop there. If you could split it up, it'd be much easier to handle

**Comment by lukeburpee** [comment]
Yep, I'll split it up and open new pull requests. @j2L4e as a side note, the resolveJsonModule flag throws a bunch of linting errors and an error regarding the implementation of the App interface ("hooks object missing from Application" and "incompatible implementation of set method"). After a little online research, the linting errors are common and are typically solved by excluding files with the json extension in the tslint config.

**Comment by NickBolles** [comment]
@lukeburpee @j2L4e I'm actually second guessing the resolve JSON issue. I think we should not resolve JSON and instead move the config into the hook, where we have access to the app, and can get all of the config from the app. Then i don't think there is any reason to have `resolveJsonModule`.

See this GIST for my current [seed-data.ts](https://gist.github.com/NickBolles/d24d4149c42b8d562a5b1992d0ad0f16)

**Comment by eddyystop** [comment]
What id the diff between your gist and the generated code?

**Comment by j2L4e** [comment]
It uses `import { readJsonFileSync } from '@feathers-plus/test-utils';` rather than `import`. This way we could leave `resolveJsonModule` turned off which has its own issues.

**Comment by eddyystop** [comment]
What is the important reason not to require config/default.js?

Its required in seed-data.?s, authentication.base.test.?s, authentication.services.test.?s, service.client.test.?s and service.server.test.?s. You will eventually argue to make changes there too if such a change is made here.

I should explain the time it takes to make such changes. The generator so far has taken over 1.4 man-years of work. The smallest change in the generated code, like adding a `// !code xxx // !end` takes at least 1 hour because that change has to be propagated through the extensive test suite.

That does not include checking the resulting code will even start up, let alone function correctly. The test suite can optionally run the generated tests automatically to test startup but that takes over 8 hours to run because of dependency loading.

The major changes suggested to seed-data.?s will take several days to implement and then test in the wild in both .js and .ts.

What is the reason justifying this effort?

**Comment by j2L4e** [comment]
"... what would be your aversion to using this json resolution method?"

It's hacky and a non-hacky alternative exists. But given the fact that `resolveJsonModule` comes with a bunch of other issues that can't be resolved easily, I'm totally fine with `declare module '*.json';`

**Comment by NickBolles** [comment]
Personally it seems like we should load the config from the app as much as possible, this allows node config to apply overrides and defualts from the multiple different config files. I think it's a mistake to load the JSON file directly. It ties the functionality to that json file directly instead of any configuration set on the app instance.

**Comment by lukeburpee** [comment]
Closing in leu of smaller pull requests

**Comment by eddyystop** [comment]
I agree with @NickBolles that app.get() should be preferred to read any of the config files.

**Pull request closed without merging**